Firestore/core/test/firebase/firestore/local/leveldb_key_test.cc

```diff
@@ -0,0 +1,363 @@
+ * description.
+ *
+ * @param key A StringView of a textual key
+ * @param key An NSString that [LevelDbKey descriptionForKey:] is expected to
+ * produce.
+ */
+#define AssertExpectedKeyDescription(expected_description, key) \
+  ASSERT_EQ((expected_description), Describe(key))
+
+TEST(LevelDbMutationKeyTest, Prefixing) {
+  auto tableKey = LevelDbMutationKey::KeyPrefix();
+  auto emptyUserKey = LevelDbMutationKey::KeyPrefix("");
+  auto fooUserKey = LevelDbMutationKey::KeyPrefix("foo");
+
+  auto foo2Key = LevelDbMutationKey::Key("foo", 2);
+
+  ASSERT_TRUE(util::StartsWith(emptyUserKey, tableKey));
+
+  // This is critical: prefixes of the a value don't convert into prefixes of
+  // the key.
+  ASSERT_TRUE(util::StartsWith(fooUserKey, tableKey));
+  ASSERT_FALSE(util::StartsWith(fooUserKey, emptyUserKey));
+
+  // However whole segments in common are prefixes.
+  ASSERT_TRUE(util::StartsWith(foo2Key, tableKey));
+  ASSERT_TRUE(util::StartsWith(foo2Key, fooUserKey));
+}
+
+TEST(LevelDbMutationKeyTest, EncodeDecodeCycle) {
+  LevelDbMutationKey key;
+  std::string user("foo");
+
+  std::vector<BatchId> batch_ids{0, 1, 100, INT_MAX - 1, INT_MAX};
+  for (auto batch_id : batch_ids) {
+    auto encoded = LevelDbMutationKey::Key(user, batch_id);
+
+    bool ok = key.Decode(encoded);
+    ASSERT_TRUE(ok);
+    ASSERT_EQ(user, key.user_id());
+    ASSERT_EQ(batch_id, key.batch_id());
+  }
+}
+
+TEST(LevelDbMutationKeyTest, Description) {
+  AssertExpectedKeyDescription("[mutation: incomplete key]",
+                               LevelDbMutationKey::KeyPrefix());
+
+  AssertExpectedKeyDescription("[mutation: user_id=user1 incomplete key]",
+                               LevelDbMutationKey::KeyPrefix("user1"));
+
+  auto key = LevelDbMutationKey::Key("user1", 42);
+  AssertExpectedKeyDescription("[mutation: user_id=user1 batch_id=42]", key);
+
+  AssertExpectedKeyDescription(
+      "[mutation: user_id=user1 batch_id=42 invalid "
+      "key=<hW11dGF0aW9uAAGNdXNlcjEAAYqqgCBleHRyYQ==>]",
+      key + " extra");
+
+  // Truncate the key so that it's missing its terminator.
+  key.resize(key.size() - 1);
+  AssertExpectedKeyDescription(
+      "[mutation: user_id=user1 batch_id=42 incomplete key]", key);
+}
+
+TEST(LevelDbDocumentMutationKeyTest, Prefixing) {
+  auto table_key = LevelDbDocumentMutationKey::KeyPrefix();
+  auto empty_user_key = LevelDbDocumentMutationKey::KeyPrefix("");
+  auto foo_user_key = LevelDbDocumentMutationKey::KeyPrefix("foo");
+
+  DocumentKey document_key = testutil::Key("foo/bar");
+  auto foo2_key = LevelDbDocumentMutationKey::Key("foo", document_key, 2);
+
+  ASSERT_TRUE(util::StartsWith(empty_user_key, table_key));
+
+  // While we want a key with whole segments in common be considered a prefix
+  // it's vital that partial segments in common not be prefixes.
+  ASSERT_TRUE(util::StartsWith(foo_user_key, table_key));
+
+  // Here even though "" is a prefix of "foo" that prefix is within a segment so
```

Comment:
Nit: commas before "that" and before "so".

Reply:
Done.